### PR TITLE
Error when required members are not set

### PR DESCRIPTION
**Pull request by 333fred**
Implements reading the required members list of a type, and enforces that required members are all set by an initializer on the constructor. Required members must be initialized with values, not with nested object initializers.

Test plan https://github.com/dotnet/roslyn/issues/57046.
Specification https://github.com/dotnet/csharplang/blob/main/proposals/required-members.md

### CODE REVIEW

**Comment by 333fred** [comment]
@jcouv @RikkiGibson please take a look.

**Comment by 333fred** [comment]
As a general note, I scoured the codebase for other places where `new BoundObjectCreation` occurs, and everything else is invocations of the `Nullable` constructor. I don't believe we care about `Nullable<T>` being defined with required members, so I didn't bother to insert checks elsewhere for it.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -5427,7 +5499,9 @@ private bool IsConstructorAccessible(MethodSymbol constructor, ref CompoundUseSi
                     type,
                     hasError);
 
-                return result;
+                CheckRequiredMembersInObjectInitializer(creation, diagnostics);
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5502]: Something similar needed for attributes?

```csharp
[My] // should be error.
[My(P = 0)] // should be okay.
public class MyAttribute : System.Attribute
{
    public required int P { get; set; }
}
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 2]
```diff
@@ -1222,4 +1222,1388 @@ struct S3
...
+    {
+        var c = @"
+public class C
+{
+    public required int Prop1 { get; set; }
+    public required int Field1;
+}
+";
+
+        var creationWithoutAssignment = @"
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1322]: variable name looks inaccurate here.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 3]
```diff
@@ -2400,6 +2400,20 @@ private void CheckForRequiredMemberAttribute(BindingDiagnosticBag diagnostics)
...
+            {
+                foreach (var member in GetMembersUnordered())
+                {
+                    if (member is not MethodSymbol method || !method.ShouldCheckRequiredMembers())
+                    {
+                        continue;
+                    }
+
+                    // The required members list for the base type '{0}' is malformed and cannot be interpreted. To use this constructor, apply the `SetsRequiredMembers` attribute.
+                    diagnostics.Add(ErrorCode.ERR_RequiredMembersBaseTypeInvalid, method.Locations[0], BaseTypeNoUseSiteDiagnostics);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2414]: Should we 'break' after this?

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 4]
```diff
@@ -1222,4 +1222,1388 @@ struct S3
...
+        CompileAndVerify(originalComp).VerifyDiagnostics();
+
+        // This IL is the equivalent of:
+        // public record Derived : Base
+        // {
+        //    public new required int P { get; init; }
+        // }
+
+        var ilSource = @"
+.assembly extern original {}
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:2219]: Do we need to use a record to meaningfully test the scenario here? Seems like IL would be much smaller without the record members.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 5]
```diff
@@ -7007,6 +7007,18 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
...
+    <value>Required member '{0}' must be set in the object initializer.</value>
+  </data>
+  <data name="ERR_RequiredMembersMustBeAssignment" xml:space="preserve">
+    <value>Required member '{0}' must be assigned a value, it cannot use a nested member or collection initializer.</value>
+  </data>
+  <data name="ERR_RequiredMembersInvalid" xml:space="preserve">
+    <value>The required members list for '{0}' is malformed and cannot be interpreted.</value>
+  </data>
+  <data name="ERR_RequiredMembersBaseTypeInvalid" xml:space="preserve">
+    <value>The required members list for the base type '{0}' is malformed and cannot be interpreted. To use this constructor, apply the `SetsRequiredMembers` attribute.</value>
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:7020]: > apply the `SetsRequiredMembers` attribute

Most diagnostic messages use `'`, not `` ` ``

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:2219]: Yes. The record is the point of this test.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2414]: No. I'm intentionally reporting on every constructor.

**Comment by 333fred** [comment]
@RikkiGibson addressed your feedback. Please take another look.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 6]
```diff
@@ -2604,6 +2573,205 @@ class C : I
...
+        );
+    }
+
+    [Fact]
+    public void RequiredMemberInAttribute_NotSet()
+    {
+        var code = @"
+using System;
+
+[Attr]
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:2589]: Consider updating the feature proposal to clearly identify this scenario as being supported by the feature.

**Review by RikkiGibson — APPROVED** [review]

**Comment by 333fred** [comment]
@jcouv for a second review.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 7]
```diff
@@ -1222,4 +1223,1555 @@ struct S3
...
+            //     public required readonly int Field1;
+            Diagnostic(ErrorCode.ERR_RequiredMemberMustBeSettable, "Field1").WithArguments("C.Field1").WithLocation(8, 34)
+        );
+    }
+
+    [Fact]
+    public void EnforcedRequiredMembers_NoInheritance_DisallowedNestedObjectInitializer()
+    {
+        var c = @"
+var c = new C() { D1 = { NestedProp = 1 }, D2 = { NestedProp = 2 } };
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1375]: nit: consider adding the version that works for contrast: `var c = new C() { D1 = new D() { NestedProp = 1 }, D2 = new D() { NestedProp = 2 } };`

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 8]
```diff
@@ -1222,4 +1223,1555 @@ struct S3
...
+    }
+
+    [Fact]
+    public void EnforcedRequiredMembers_ShadowedInSource_01()
+    {
+        var vb = @"
+Imports System.Runtime.CompilerServices
+<RequiredMember>
+Public Class Base
+    <RequiredMember>
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1820]: Are we going to disallow usage of this attribute in VB?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 9]
```diff
@@ -5747,7 +5842,23 @@ private BoundExpression BindTypeParameterCreationExpression(ObjectCreationExpres
                 }
             }
 
-            diagnostics.Add(errorLocation, useSiteInfo);
+            if (suppressUnsupportedRequiredMembersError && useSiteInfo.AccumulatesDiagnostics && useSiteInfo.Diagnostics is { Count: not 0 })
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5845]: I didn't understand the benefit of this suppression. If a constructor gets picked and it has a problem, why not report it?

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 8 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1820]: I wasn't planning on it.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 9 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5845]: Because then you'll get the error, even if you decorate the constructor with `SetsRequiredMembers` (which will be supported in a later pr.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 10]
```diff
@@ -4967,6 +4968,96 @@ private static void ReportDuplicateObjectMemberInitializers(BoundExpression boun
...
+
+            if (constructor.ContainingType.HasRequiredMembersError)
+            {
+                // An error will be reported on the constructor if from source, or a use-site diagnostic will be reported on the use if from metadata.
+                return;
+            }
+
+            var requiredMembers = constructor.ContainingType.AllRequiredMembers.ToBuilder();
+
+            if (requiredMembers.Count == 0)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:4991]: nit: Consider moving count check before `ToBuilder()` allocation

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 11]
```diff
@@ -4967,6 +4968,96 @@ private static void ReportDuplicateObjectMemberInitializers(BoundExpression boun
...
+            reportMembers();
+
+            void reportMembers()
+            {
+                Location location = creationSyntax switch
+                {
+                    ObjectCreationExpressionSyntax { Type: { } type } => type.Location,
+                    BaseObjectCreationExpressionSyntax { NewKeyword: { } newKeyword } => newKeyword.GetLocation(),
+                    AttributeSyntax { Name: { } name } => name.Location,
+                    _ => creationSyntax.Location
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5049]: Can we hit this arm?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 12]
```diff
@@ -4967,6 +4968,96 @@ private static void ReportDuplicateObjectMemberInitializers(BoundExpression boun
...
+                    ? (assignmentOperator.Left switch
+                    {
+                        // Regular initializers
+                        BoundObjectInitializerMember member => member.MemberSymbol,
+                        // Attribute initializers
+                        BoundPropertyAccess propertyAccess => propertyAccess.PropertySymbol,
+                        BoundFieldAccess fieldAccess => fieldAccess.FieldSymbol,
+                        _ => null
+                    }, assignmentOperator.Right)
+                    : default;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5014]: nit: expression is hard to parse. Consider just grabbing the memberSymbol here, then grabbing the right part later on when we need it.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 9 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5845]: Thanks. Consider leaving a comment to that effect

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 13]
```diff
@@ -7007,6 +7007,18 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_RequiredMemberMustBeSettable" xml:space="preserve">
     <value>Required member '{0}' must be settable.</value>
   </data>
+  <data name="ERR_RequiredMemberMustBeSet" xml:space="preserve">
+    <value>Required member '{0}' must be set in the object initializer or attribute constructor.</value>
+  </data>
+  <data name="ERR_RequiredMembersMustBeAssignment" xml:space="preserve">
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:7013]: >MustBeAssignment [](http://example.com/codeflow?start=33&length=16)

nit: "MustBeAssignment" sounds off.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 2). Will look at tests this afternoon

File: src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs [hunk 14]
```diff
@@ -499,6 +509,118 @@ public override string MetadataName
...
+                _lazyHasRequiredMembersErrorDoNotAccessDirectly = !success;
+                // InterlockedInitialize uses `Interlocked.CompareExchange` under the hood, which will introduce a full
+                // memory barrier, ensuring that the _lazyHasRequiredMembersErrorDoNotAccessDirectly write is not reordered
+                // after the initialization of _lazyRequiredMembers.
+                RoslynImmutableInterlocked.InterlockedInitialize(ref _lazyRequiredMembers, requiredMembers);
+            }
+            else
+            {
+                // Ensure that, after this method, _lazyRequiredMembersErrorDoNotAccessDirectly is able to be accessed safely.
+                Interlocked.MemoryBarrier();
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:563]: `_lazyRequiredMembers` is a struct. I can't really make a "sentinel" value with it, and thus can't use it to represent what I need.

[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:563]: Thanks. I was looking at the wrong `ImmutableSegmentedDictionary` (non-generic one is a class).
Check to see what Aleksey thinks, but I would rather wrap the struct and stay in a less error-prone concurrency pattern.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1857]: Added a note to test plan

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 15]
```diff
@@ -1222,4 +1223,1592 @@ struct S3
...
+    [Fact]
+    public void EnforcedRequiredMembers_NoInheritance_Unsettable()
+    {
+        var c = @"
+var c = new C();
+c = new C() { Prop1 = 1, Field1 = 1 };
+
+public class C
+{
+    public required int Prop1 { get; }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1344]: Do we have a test with required unsettable property from metadata?

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 16]
```diff
@@ -1222,4 +1223,1592 @@ struct S3
...
+{
+    public required int P { get; init; }
+}
+";
+
+        var originalComp = CreateCompilationWithRequiredMembers(new[] { original, IsExternalInitTypeDefinition }, assemblyName: "original");
+        CompileAndVerify(originalComp, verify: ExecutionConditionUtil.IsCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+
+        // This IL is the equivalent of:
+        // public record Derived : Base
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:2228]: nit: For these `CompileIL`, I don't see the point of bloating the IL by making it a record. A class would have sufficed to the purpose of those tests.

File: src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:563]: Let's discuss this offline. I think we shouldn't invent new ways of handling lazy initialization. For structs we usually declare wrapper classes in order to ensure atomic read/write operations. See, for example, ```TypeWithAnnotations.Boxed```. I probably could be convinced otherwise if this situation is unique in some way.

[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:563]: I'm ok with doing a `Boxed` struct or similar here. I was generally trying to avoid an allocation, but I'm fine with saving that for if we decide there's a perf issue.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 16 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:2228]: It would not. The point of these tests is for copy constructors and similar, and in a future PR I'm working on these being records pointed out some considerations I'll need to think about for how copy constructors work with `SetsRequiredMembers`. I'll keep these as is.

**Comment by 333fred** [comment]
@jcouv addressed feedback

@sharwell please review https://github.com/dotnet/roslyn/pull/60101/commits/31157e3121ff0b6178382317f8943a9c061e0fc8 for changes to `ImmutableSegmentedDictionary`.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 15 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1344]: Gentle ping

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 6) with one test to consider

File: src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs [hunk 17]
```diff
@@ -499,6 +509,119 @@ public override string MetadataName
...
+            RoslynImmutableInterlocked.InterlockedCompareExchange(ref _lazyRequiredMembers, requiredMembers, RequiredMembersSentinel);
+        }
+
+        /// <summary>
+        /// Attempts to calculate the required members for this type. Returns false if there were errors.
+        /// </summary>
+        private bool TryCalculateRequiredMembers(ref ImmutableSegmentedDictionary<string, Symbol>.Builder? requiredMembersBuilder)
+        {
+            var lazyRequiredMembers = _lazyRequiredMembers;
+            if (!ImmutableSegmentedDictionary.ReferenceEquals(RequiredMembersSentinel, lazyRequiredMembers))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:569]: Cool solution. Much easier to follow now. Thanks!

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs (hunk 15 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:1344]: I added one, just below this one.

File: src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs [hunk 18]
```diff
@@ -70,7 +70,18 @@ namespace Microsoft.CodeAnalysis.Collections
...
+        /// <summary>
+        /// Creates a new empty segmented dictionary, that is not underlying-reference equal to <see cref="Empty"/>. The intent is for this to
+        /// be used for sentinel values only.
+        /// </summary>
+        internal static ImmutableSegmentedDictionary<TKey, TValue> CreateEmpty() => new(new SegmentedDictionary<TKey, TValue>());
+
+        /// <summary>
+        /// Compares the given segmented dictionaries for reference-equality of their underlying storage.
+        /// </summary>
+        internal static bool ReferenceEquals(ImmutableSegmentedDictionary<TKey, TValue> left, ImmutableSegmentedDictionary<TKey, TValue> right) => ReferenceEquals(left._dictionary, right._dictionary);
```
Comment (sharwell) [inline src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs:84]: This is already implemented as `operator==`.

File: src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs [hunk 19]
```diff
@@ -70,7 +70,18 @@ namespace Microsoft.CodeAnalysis.Collections
...
         where TKey : notnull
     {
-        public static readonly ImmutableSegmentedDictionary<TKey, TValue> Empty = new(new SegmentedDictionary<TKey, TValue>());
+        public static readonly ImmutableSegmentedDictionary<TKey, TValue> Empty = CreateEmpty();
+
+        /// <summary>
+        /// Creates a new empty segmented dictionary, that is not underlying-reference equal to <see cref="Empty"/>. The intent is for this to
+        /// be used for sentinel values only.
+        /// </summary>
+        internal static ImmutableSegmentedDictionary<TKey, TValue> CreateEmpty() => new(new SegmentedDictionary<TKey, TValue>());
```
Comment (sharwell) [inline src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs:79]: 📝 This doesn't seem like a generally-valuable API. For the edge case where it's needed, it would be better to construct the sentinel instance using private reflection rather than facilitate this type of usage in other scenarios.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7)
